src/Storages/QueryCache.h

```diff
@@ -0,0 +1,229 @@
+    {
+        return ast->getTreeHash() == other.ast->getTreeHash()
+               && header.getNamesAndTypesList() == other.header.getNamesAndTypesList()
+               && settings == other.settings
+               && username == other.username;
+    }
+
+    ASTPtr ast;
+    Block header;
+    Settings settings;
+    std::optional<String> username;
+};
+
+struct CacheKeyHasher
+{
+    size_t operator()(const CacheKey & key) const
+    {
+         SipHash hash;
+         hash.update(key.ast->getTreeHash());
+         hash.update(key.header.getNamesAndTypesList().toString());
+         for (const auto & setting : key.settings)
+         {
+             hash.update(setting.getValueString());
+         }
+         if (key.username.has_value())
+         {
+             hash.update(*key.username);
+         }
+         return hash.get64();
+    }
+};
+
+struct QueryWeightFunction
+{
+    size_t operator()(const Data & data) const
+    {
+        const Block & block = data.first;
+        const Chunks & chunks = data.second;
+
+        size_t res = 0;
+        for (const auto & chunk : chunks)
+        {
+            res += chunk.allocatedBytes();
+        }
+        res += block.allocatedBytes();
+
+        return res;
+    }
+};
+
+class CacheRemovalScheduler
+{
+private:
+    using Timestamp = std::chrono::time_point<std::chrono::high_resolution_clock>;
+    using Duration = std::chrono::high_resolution_clock::duration;
+public:
+    void scheduleRemoval(Duration duration, CacheKey cache_key)
+    {
+        std::unique_lock lock(mutex);
+        TimedCacheKey timer = {now() + duration, cache_key};
+        queue.push(timer);
+        auto top = queue.top();
+        lock.unlock();
+
+        // if the newly scheduled timer turned out to have the smallest timestamp in the entire queue, notify timer_cv
+        if (top == timer)
+        {
+            timer_cv.notify_one();
+        }
+    }
+
+    template <typename Cache>
+    void processRemovalQueue(Cache * query_cache)
+    {
+        while (process_removal_queue.load())
+        {
+            std::unique_lock lock(mutex);
+
+            // take the timer with the lowest timestamp from the queue if there is one
+            const std::optional<TimedCacheKey> awaited_timer = nextTimer();
+
+            // wake up if either a timer with a lower timestamp than awaited_timer was pushed to the queue, the awaited_timer went off or the server was stoped
+            timer_cv.wait_until(lock, awaited_timer.has_value() ? awaited_timer->time : infinite_time);
+
+            // if awaited_timer went off, remove entry from cache
+            if (awaited_timer.has_value() && awaited_timer->time <= now())
+            {
+                lock.unlock();
+                queue.pop();
+                query_cache->remove(awaited_timer->cache_key);
+            }
+        }
+    }
+
+    void stopProcessingRemovalQueue()
+    {
+        process_removal_queue.store(false);
```

Comment:
Better to rename flag into `is_finished`.

Reply:
Done.